### PR TITLE
Bring PHP functionality in line with recent JS changes and add integration tests for third parties

### LINKED ISSUE #66: Add PHP test coverage for the actual third parties implemented
So far, the PHP codebase only has unit test coverage for all its foundation. While this is great, what we're missing is a set of integration tests that ensure that with the JSON file structure the output of the HTML, scripts etc. generated matches the expected.

This is particularly crucial in this repository since we're dealing with two implementations of the same thing (one in JS, one in PHP), but they always need to match, or be kept in sync.

@flashdesignory Looking at some of the PRs that landed over the recent months, I have a hunch the PHP codebase will have some bugs now because it seems some of the JSON file values were changed in technically backward-incompatible ways - which would be fine for JS because you've updated it accordingly, but I think the corresponding PHP code is now outdated.

It's a bit of a painful situation to have to keep these in sync, but having test coverage should ensure we don't miss anything in that regard in the future. Separately, I wonder if there's a better way for us to stay on track of such changes. FWIW with the PHP side we're mostly following on what you do on the JS side, so it may be good to simply flag any PR where the JSON is changed to us (@adamsilverstein and me). Let me know if you have other ideas how we can improve this. In any case, PHP test coverage for this should help.

### PR DESCRIPTION
**Pull request by felixarntz**
Fixes #66.

This PR makes several enhancements:
* Add support for `optionalParams` for scripts, as was recently introduced in the JS codebase (see #58).
* Add third party implementation for Google Tag Manager, as it was recently added in the JS codebase (see #36).
* Add PHPUnit integration test coverage to ensure the output for all third parties works as expected (see #66).
* Fix various bugs that were exposed by the new integration tests.

It may be helpful for the review process to review the individual commits as part of this PR, rather than all changes at once.

Effectively, this PR does everything needed to bring it up to date with the JS codebase, and it includes test coverage to make sure in the future it doesn't get out of date without being noticed.

### CODE REVIEW

File: inc/Data/ThirdPartyDataFormatter.php
Comment (adamsilverstein) [inline inc/Data/ThirdPartyDataFormatter.php:203]: I assume avoiding overwriting existing values is to match the JS behavior?

File: inc/Data/ThirdPartyDataFormatter.php
Comment (adamsilverstein) [inline inc/Data/ThirdPartyDataFormatter.php:241]: why does this return '""'?

File: tests/phpunit/utils/TestCase.php
Comment (adamsilverstein) [inline tests/phpunit/utils/TestCase.php:26]: Nice!

**Review by adamsilverstein — COMMENTED** [review]
Lookin' good! I left a couple of small questions.

File: inc/Data/ThirdPartyDataFormatter.php
Comment (felixarntz) [reply inc/Data/ThirdPartyDataFormatter.php:203]: Partially yes. Mostly this uses a `foreach` to have the query parameters sorted based on `params`. I initially used `array_merge($optionalArgs, $queryArgs)`, but that will lead for values from `optionalParams` to come first.

Not a real problem I assume, but it leads to different URLs than the reference snippets for e.g. Google Analytics recommend, so I thought it's worth keeping it consistent with those.

File: inc/Data/ThirdPartyDataFormatter.php
Comment (felixarntz) [reply inc/Data/ThirdPartyDataFormatter.php:241]: It's the same as `json_encode( '' )`, but since that's predictable, I just made it `'""'`. I can add a comment to clarify.

Comment (felixarntz) [reply inc/Data/ThirdPartyDataFormatter.php]: Done in https://github.com/GoogleChromeLabs/third-party-capital/pull/70/commits/58de6d48a05f4e1a5994c910d96fd59894039338

**Review by housseindjirdeh — APPROVED** [review]
Cheers thanks for doing this 😍 

Will wait for @adamsilverstein to approve before merging

**Review by adamsilverstein — APPROVED** [review]
Nice work!

**Comment by felixarntz** [comment]
Pushed one more fix here in https://github.com/GoogleChromeLabs/third-party-capital/pull/70/commits/b38ca1214a4742d20037b83c7b3d699ba9eda2e3 because the schema changes from #71 broke PHPUnit tests again.

**Pull request merged**
